### PR TITLE
bpo-45773: Stop "optimizing" certain jump patterns

### PR DESCRIPTION
**Pull request by brandtbucher**
https://bugs.python.org/issue45773

### CODE REVIEW

**Review by markshannon — CHANGES_REQUESTED** [review]
I'm not sure about the comments `// It's okay if inst->i_target == target->i_target.`, or the asserts `assert(inst->i_target != target->i_target);`

I'm not saying that they are incorrect, but it is not obvious why they are true.
Could you elaborate?

Something else to consider is to create a `jump_thread` function which gets all the checks if one place (different targets, compatible line numbers). It should return 1 if successful.

We could then replace code from line 8253 to 8256 with `i -= jump_thread(inst, target);`

**Comment by brandtbucher** [comment]
I figured I'd give your `jump_thread` suggestion a stab, and that sort of led to a refactoring of all jump elimination. I do really like the way the logic here cleans up as a result.

If you think this is outside the scope of this PR, though, I can revert the latest changes and propose them separately.

File: Python/compile.c [hunk 1]
```diff
@@ -8250,23 +8245,19 @@ optimize_basic_block(struct compiler *c, basicblock *bb, PyObject *consts)
...
-                        }
+                        i -= jump_thread(inst, target, JUMP_IF_FALSE_OR_POP);
                         break;
                     case JUMP_IF_TRUE_OR_POP:
-                        assert (inst->i_target->b_iused == 1);
+                    case POP_JUMP_IF_TRUE:
                         if (inst->i_lineno == target->i_lineno) {
+                            // We don't need to bother checking for loops here,
+                            // since inst->i_target can't possibly be equal to
+                            // inst->i_target->b_next:
```
Comment (markshannon) [inline Python/compile.c]: Why can't it be equal?
Could you use `jump_thread` here as well? The redundant check is harmless.

File: Python/compile.c [hunk 2]
```diff
@@ -8278,23 +8269,19 @@ optimize_basic_block(struct compiler *c, basicblock *bb, PyObject *consts)
...
-                        }
+                        i -= jump_thread(inst, target, JUMP_IF_TRUE_OR_POP);
                         break;
                     case JUMP_IF_FALSE_OR_POP:
-                        assert (inst->i_target->b_iused == 1);
+                    case POP_JUMP_IF_FALSE:
                         if (inst->i_lineno == target->i_lineno) {
+                            // We don't need to bother checking for loops here,
+                            // since inst->i_target can't possibly be equal to
+                            // inst->i_target->b_next.
```
Comment (markshannon) [inline Python/compile.c]: Likewise

**Comment by markshannon** [comment]
> I figured I'd give your `jump_thread` suggestion a stab, and that sort of led to a refactoring of all jump elimination. I do really like the way the logic here cleans up as a result.

I agree, this does clean things up.

File: Python/compile.c (hunk 1 above)
Comment (brandtbucher) [reply Python/compile.c]: We don’t want to use `jump_thread` here. Unlike all the others, these cases don’t actually take the second jump; they “fall through” to the block following their target (notice that `inst->i_target` becomes `inst->i_target->b_next`, not `target->i_target`). 

As to the need to check for loops: I don’t see how a block could possibly fall through to itself. Should I add an assert, just to be sure?

Comment (markshannon) [reply Python/compile.c]: `b_next`  shouldn't point to itself (I think if it did we would have an infinite loop somewhere).
Maybe change the comment from
"inst->i_target can't possibly be equal to inst->i_target->b_next" to
"a block's b_next cannot point to itself".
An assertion would do no harm.

**Pull request merged**